xlators/storage/posix/src/posix-helpers.c

```diff
@@ -482,21 +479,21 @@ _posix_xattr_get_set(dict_t *xattr_req, char *key, data_t *data,
 
             ret = sys_close(_fd);
             _fd = -1;
-            if (ret == -1) {
+            if (ret != 0) {
                 gf_msg(filler->this->name, GF_LOG_ERROR, errno,
                        P_MSG_XDATA_GETXATTR, "Close on file %s failed",
                        filler->real_path);
                 goto err;
             }
-
-            ret = dict_set_bin(filler->xattr, key, databuf,
-                               filler->stbuf->ia_size);
+        set_content_key_in_dict:
+            ret = dict_set_bin(filler->xattr, GF_CONTENT_KEY, databuf,
```

Comment:
`dict_set_bin()` doesn't accept a NULL buffer, even if it's of size 0. We should initialize `databuf` with something safe, like a constant empty string for example. It will be accessible for the whole life of the dict (the value should be quite irrelevant for a buffer of size 0, but using it we make sure that the pointer will remain valid in case it's dereferenced for some reason).

Reply:
This whole patch started when I wanted to remove the useless CALLOC() of a buffer that will be read into, and turn it into a MALLOC(). If filler->stbuf->ia_size is zero, what would the result of GF_MALLOC/GF_CALLOC would have been? Is that a theoretical case we are discussing here?

Reply:
I don't think it's a theoretical thing. While the specification says that this case is undefined behavior, actually the implementations of `malloc()` in both glibc and tcmalloc return a good pointer even if size is 0, and I don't see anything in GF_MALLOC() that changes this behavior.

I would keep the same behavior.

Reply:
ACK.

Reply:
I think we didn't understood each other. I was ok with the approach of avoiding some system calls if the file is empty. The only problem was that we still need to put something of size 0 into the dictionary. I think that just initializing `databuf` with a static empty string would be enough to avoid the issue and it's safe:

```c
    char *databuf = "";
```

If the file is not empty, this value will be overridden by the allocated memory. Otherwise it will be used to add a zero-size value to the dict without using a NULL pointer, which is not allowed.